File: app/src/main/java/com/ty/zbpet/ui/activity/product/ProductTodoDetailActivity.java
package com.ty.zbpet.ui.activity.product;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.util.SparseArray;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.bigkoo.pickerview.listener.OnTimeSelectListener;
import com.ty.zbpet.R;
import com.ty.zbpet.bean.ResponseInfo;
import com.ty.zbpet.bean.UserInfo;
import com.ty.zbpet.bean.product.ProductTodoDetails;
import com.ty.zbpet.bean.product.ProductTodoSave;
import com.ty.zbpet.net.HttpMethods;
import com.ty.zbpet.presenter.product.ProducePresenter;
import com.ty.zbpet.presenter.product.ProductUiObjInterface;
import com.ty.zbpet.ui.activity.ScanBoxCodeActivity;
import com.ty.zbpet.ui.adapter.product.ProductTodoDetailAdapter;
import com.ty.zbpet.ui.base.BaseActivity;
import com.ty.zbpet.ui.widght.SpaceItemDecoration;
import com.ty.zbpet.util.CodeConstant;
import com.ty.zbpet.util.DataUtils;
import com.ty.zbpet.util.ResourceUtil;
import com.ty.zbpet.util.ZBLog;
import com.ty.zbpet.util.ZBUiUtils;
import com.zhouyou.http.exception.ApiException;
import com.zhouyou.http.subsciber.BaseSubscriber;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import okhttp3.RequestBody;

/**
 * 生产入库 待办详情
 *
 * @author TY
 */
public class ProductTodoDetailActivity extends BaseActivity implements ProductUiObjInterface<ProductTodoDetails>
        , ProductTodoDetailAdapter.SaveEditListener {


    private RecyclerView reView;
    private TextView tvTime;
    private TextView backGoods;
    private TextView selectHouse;
    private TextView tvPath;
    private TextView tvType;
    private EditText etDesc;

    private ProductTodoDetailAdapter adapter;

    private String selectTime;
    private String sapOrderNo;

    private List<ProductTodoDetails.ListBean> oldList = new ArrayList<>();
    private List<ProductTodoDetails.WarehouseListBean> houseList = new ArrayList<>();

    private final static int REQUEST_SCAN_CODE = 1;
    private final static int RESULT_SCAN_CODE = 2;

    private ProducePresenter presenter = new ProducePresenter(this);

    /**
     * 用户信息
     */
    private UserInfo userInfo;

    /**
     * 保存用户在输入框中的数据
     */
    private SparseArray<String> numberArray = new SparseArray(10);
    private SparseArray<String> contentArray = new SparseArray(10);
    private SparseArray<String> startCodeArray = new SparseArray(10);
    private SparseArray<String> endCodeArray = new SparseArray(10);
    private SparseArray<String> sapArray = new SparseArray(10);
    private SparseArray<ArrayList<String>> carCodeArray = new SparseArray(10);
    /**
     * 库位码 ID
     */
    private SparseArray<String> positionId = new SparseArray(10);

    /**
     * 仓库 ID
     */
    private String warehouseId;

    /**
     * 列表 ID
     */
    private int itemId = -1;

    /**
     * 箱码
     */
    private ArrayList<String> boxCodeList = new ArrayList<>();


    @Override
    protected void onBaseCreate(Bundle savedInstanceState) {

    }

    @Override
    protected int getActivityLayout() {
        return R.layout.activity_content_row_three;
    }

    @Override
    protected void initOneData() {
        sapOrderNo = getIntent().getStringExtra("sapOrderNo");

        presenter.fetchProductTodoInfo(sapOrderNo);
    }

    @Override
    protected void initTwoView() {

        // in_storage_detail 到货明细

        initToolBar(R.string.label_produce_in_storage, new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                productTodoSave(initTodoBody());
            }
        });

        reView = findViewById(R.id.rv_in_storage_detail);
        tvTime = findViewById(R.id.tv_time);
        backGoods = findViewById(R.id.in_storage_detail);
        selectHouse = findViewById(R.id.tv_house);

        tvPath = findViewById(R.id.tv_path);
        tvType = findViewById(R.id.tv_type);
        etDesc = findViewById(R.id.et_desc);

        SimpleDateFormat format = new SimpleDateFormat(CodeConstant.DATE_SIMPLE_H_M, Locale.CHINA);
        selectTime = format.format(new Date());

        tvTime.setText(selectTime);
        backGoods.setText("入库明细");


        tvTime.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ZBUiUtils.showPickDate(v.getContext(), new OnTimeSelectListener() {
                    @Override
                    public void onTimeSelect(Date date, View v) {
                        selectTime = ZBUiUtils.getTime(date);
                        tvTime.setText(selectTime);

                        ZBUiUtils.showToast(selectTime);
                    }
                });
            }
        });

        // 仓库默认值设置
        DataUtils.setHouseId(0, 0);

        userInfo = DataUtils.getUserInfo();
        List<UserInfo.WarehouseListBean> warehouseList = userInfo.getWarehouseList();

        int size = warehouseList.size();
        final ArrayList<String> houseName = new ArrayList<>();
        for (int i = 0; i < size; i++) {
            houseName.add(warehouseList.get(i).getWarehouseName());
        }
        selectHouse.setText(houseName.get(0));

        selectHouse.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                ZBUiUtils.selectDialog(v.getContext(), 0, houseName, selectHouse);
            }
        });

    }

    /**
     * 出库 保存
     *
     * @param body 参数
     */
    private void productTodoSave(RequestBody body) {

        if (body == null) {
            return;
        }

        HttpMethods.getInstance().getProduceTodoSave(new BaseSubscriber<ResponseInfo>() {
            @Override
            public void onError(ApiException e) {
                ZBUiUtils.showToast(e.getMessage());
            }

            @Override
            public void onNext(ResponseInfo responseInfo) {
                if (CodeConstant.SERVICE_SUCCESS.equals(responseInfo.getTag())) {
                    // 入库成功（保存）
                    ZBUiUtils.showToast(responseInfo.getMessage());
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            finish();
                        }
                    });
                } else {
                    ZBUiUtils.showToast(responseInfo.getMessage());
                }
            }
        }, body);
    }

    /**
     * 构建 保存 的 Body
     *
     * @return
     */
    private RequestBody initTodoBody() {

        ProductTodoSave requestBody = new ProductTodoSave();
        List<ProductTodoSave.DetailsBean> detail = new ArrayList<>();

        SparseArray<Integer> houseId = DataUtils.getHouseId();

        List<UserInfo.WarehouseListBean> warehouseList = userInfo.getWarehouseList();

        // 仓库信息
        // 仓库 ID 扫箱码获取
        // warehouseId = warehouseList.get(integer).getWarehouseId();
        String warehouseNo;
        String warehouseName;
        if (houseId == null) {
            warehouseNo = warehouseList.get(0).getWarehouseNo();
            warehouseName = warehouseList.get(0).getWarehouseName();
        } else {
            Integer integer = houseId.get(0);
            warehouseNo = warehouseList.get(integer).getWarehouseNo();
            warehouseName = warehouseList.get(integer).getWarehouseName();
        }

        int size = oldList.size();
        for (int i = 0; i < size; i++) {
            List<String> boxQrCode = carCodeArray.get(i);
            String bulkNum = numberArray.get(i);
            String content = contentArray.get(i);
            String startCode = startCodeArray.get(i);
            String endCode = endCodeArray.get(i);
            String sap = sapArray.get(i);

            String Id = positionId.get(i);

            ProductTodoSave.DetailsBean bean = new ProductTodoSave.DetailsBean();
            if (!TextUtils.isEmpty(bulkNum) && boxQrCode.size() > 0) {

                String goodsId = oldList.get(i).getGoodsId();

                bean.setPositionId(Id);
                bean.setNumber(bulkNum);
                bean.setContent(content);

                bean.setSapMaterialBatchNo(sap);
                bean.setStartQrCode(startCode);
                bean.setEndQrCode(endCode);

                bean.setGoodsId(goodsId);
                bean.setBoxQrCode(boxQrCode);


                detail.add(bean);
            } else {
                // 跳出当前一列、不处理
                continue;
            }
        }
        // 没有合法的操作数据,不请求网络
        if (detail.size() == 0) {
            ZBUiUtils.showToast("请完善您要保存的信息");
            return null;
        }

        String remark = etDesc.getText().toString().trim();
        String time = tvTime.getText().toString().trim();

        requestBody.setDetails(detail);

        requestBody.setWarehouseId(warehouseId);
        requestBody.setWarehouseNo(warehouseNo);
        requestBody.setWarehouseName(warehouseName);

        requestBody.setSapOrderNo(sapOrderNo);
        requestBody.setInTime(time);
        requestBody.setRemark(remark);


        String json = DataUtils.toJson(requestBody, 1);
        ZBLog.e("JSON " + json);
        return RequestBody.create(okhttp3.MediaType.parse("application/json;charset=UTF-8"), json);
    }


    @Override
    public void detailObjData(ProductTodoDetails details) {

        oldList = details.getList();

        if (adapter == null) {
            LinearLayoutManager manager = new LinearLayoutManager(ResourceUtil.getContext());
            reView.addItemDecoration(new SpaceItemDecoration(ResourceUtil.dip2px(10), false));
            reView.setLayoutManager(manager);
            adapter = new ProductTodoDetailAdapter(this, R.layout.item_produce_detail_todo, oldList);
            reView.setAdapter(adapter);

            adapter.setOnItemClickListener(new ProductTodoDetailAdapter.OnItemClickListener() {
                @Override
                public void onItemClick(View view, RecyclerView.ViewHolder holder, final int position) {

                    View rlDetail = holder.itemView.findViewById(R.id.rl_detail);
                    ImageView ivArrow = holder.itemView.findViewById(R.id.iv_arrow);
                    Button bindingCode = holder.itemView.findViewById(R.id.btn_binding_code);

                    if (rlDetail.getVisibility() == View.VISIBLE) {
                        rlDetail.setVisibility(View.GONE);
                        ivArrow.setImageResource(R.mipmap.ic_collapse);
                    } else {
                        rlDetail.setVisibility(View.VISIBLE);
                        ivArrow.setImageResource(R.mipmap.ic_expand);
                    }

                    bindingCode.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            itemId = position;
                            Intent intent = new Intent(ProductTodoDetailActivity.this, ScanBoxCodeActivity.class);
                            intent.putExtra("itemId", itemId);
                            intent.putExtra(CodeConstant.PAGE_STATE, true);
                            intent.putStringArrayListExtra("boxCodeList", carCodeArray.get(itemId));
                            startActivityForResult(intent, REQUEST_SCAN_CODE);
                        }
                    });

                    ZBUiUtils.hideInputWindow(view.getContext(), view);

                }

                @Override
                public boolean onItemLongClick(View view, RecyclerView.ViewHolder holder, int position) {
                    return false;
                }
            });
        } else {
            adapter.notifyDataSetChanged();
        }

    }


    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_SCAN_CODE && resultCode == RESULT_SCAN_CODE) {
            itemId = data.getIntExtra("itemId", -1);
            warehouseId = data.getStringExtra("warehouseId");
            boxCodeList = data.getStringArrayListExtra("boxCodeList");
            carCodeArray.put(itemId, boxCodeList);
        }
    }

    @Override
    public void saveEditAndGetHasFocusPosition(String etType, Boolean hasFocus, int position, EditText editText) {

        String textContent = editText.getText().toString().trim();

        if (CodeConstant.ET_NUMBER.equals(etType)) {
            numberArray.put(position, textContent);
        } else if (CodeConstant.ET_CONTENT.equals(etType)) {
            contentArray.put(position, textContent);
        } else if (CodeConstant.ET_BATCH_NO.equals(etType)) {
            sapArray.put(position, textContent);
        } else if (CodeConstant.ET_START_CODE.equals(etType)) {
            startCodeArray.put(position, textContent);
        } else if (CodeConstant.ET_END_CODE.equals(etType)) {
            endCodeArray.put(position, textContent);
        }

    }
}

File: app/src/main/java/com/ty/zbpet/ui/activity/product/ProductTodoDetailActivity.kt
package com.ty.zbpet.ui.activity.product

import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.util.SparseArray
import android.view.View
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import com.ty.zbpet.R
import com.ty.zbpet.bean.UserInfo
import com.ty.zbpet.bean.product.ProductDetails
import com.ty.zbpet.bean.product.ProductTodoSave
import com.ty.zbpet.constant.CodeConstant
import com.ty.zbpet.data.SharedP
import com.ty.zbpet.net.RequestBodyJson
import com.ty.zbpet.presenter.product.ProducePresenter
import com.ty.zbpet.presenter.product.ProductUiListInterface
import com.ty.zbpet.ui.activity.ScanBoxCodeActivity
import com.ty.zbpet.ui.adapter.product.ProductTodoDetailAdapter
import com.ty.zbpet.ui.base.BaseActivity
import com.ty.zbpet.ui.widght.ShowDialog
import com.ty.zbpet.ui.widght.SpaceItemDecoration
import com.ty.zbpet.util.*
import com.xiasuhuei321.loadingdialog.view.LoadingDialog
import com.zhy.adapter.recyclerview.MultiItemTypeAdapter
import kotlinx.android.synthetic.main.activity_content_row_three.*
import okhttp3.RequestBody
import java.text.SimpleDateFormat
import java.util.*

/**
 * 生产入库 待办详情
 *
 * @author TY
 */
class ProductTodoDetailActivity : BaseActivity()
        , ProductUiListInterface<ProductDetails.ListBean> {

    private var adapter: ProductTodoDetailAdapter? = null

    private var selectTime: String? = null
    private var sapOrderNo: String? = null
    private var sapFirmNo: String? = null
    private var content: String = ""
    private var sign: String = ""

    private var oldList: List<ProductDetails.ListBean> = ArrayList()

    private val presenter = ProducePresenter(this)

    /**
     * 用户信息
     */
    private lateinit var userInfo: UserInfo
    private lateinit var warehouseList: MutableList<UserInfo.WarehouseListBean>

    /**
     * 保存箱码的数据
     */
    private val carCodeArray: SparseArray<ArrayList<String>> = SparseArray(10)
    /**
     * 库位码 ID
     */
    private val positionId: SparseArray<String> = SparseArray(10)


    /**
     * 列表 ID
     */
    private var itemId = -1

    /**
     * 箱码
     */
    private var boxCodeList = ArrayList<String>()


    /**
     * 仓库 name
     */
    private val houseName = ArrayList<String>()

    override val activityLayout: Int
        get() = R.layout.activity_content_row_three

    override fun onBaseCreate(savedInstanceState: Bundle?) {

    }

    override fun initOneData() {
        sapOrderNo = intent.getStringExtra("sapOrderNo")
        sapFirmNo = intent.getStringExtra("sapFirmNo")
        content = intent.getStringExtra("content")
        sign = intent.getStringExtra("sign")

        // 仓库默认值设置
//        DataUtils.setHouseId(0, 0)
        SharedP.putWarehouseId(this@ProductTodoDetailActivity, 0)

        // 不登录用户数据：DataUtils.getUserInfo()
        userInfo = SimpleCache.getUserInfo(CodeConstant.USER_DATA)
        warehouseList = userInfo.warehouseList

        val size = warehouseList.size
        for (i in 0 until size) {
            houseName.add(warehouseList[i].warehouseName.toString())
        }
        //tv_house!!.text = houseName[0]
        try {
            tv_house.text = warehouseList[0].warehouseName
        } catch (e: Exception) {
            e.printStackTrace()
        }

        presenter.fetchProductTodoInfo(sign, sapOrderNo)
    }

    override fun initTwoView() {

        initToolBar(R.string.label_produce_in_storage_detail, "保存", View.OnClickListener { productTodoSave(initTodoBody()) })

        val format = SimpleDateFormat(CodeConstant.DATE_SIMPLE_H_M, Locale.CHINA)
        selectTime = format.format(Date())

        tv_time.text = selectTime
        in_storage_detail.text = "入库明细"

        tv_time.setOnClickListener { v ->
            ZBUiUtils.showPickDate(v.context) { date, _ ->
                selectTime = ZBUiUtils.getTime(date)
                tv_time.text = selectTime

                ZBUiUtils.showSuccess(selectTime)
            }
        }

        // 用户选择仓库信息
        tv_house!!.setOnClickListener { v -> ZBUiUtils.selectDialog(v.context, CodeConstant.SELECT_HOUSE_BUY_IN, 0, houseName, tv_house) }

    }

    /**
     * 出库 保存
     *
     * @param body 参数
     */
    private fun productTodoSave(body: RequestBody?) {

        if (body == null) {
            return
        }
        presenter.productTodoSave(body)

    }

    /**
     * 构建 保存 的 Body
     *
     * @return
     */
    private fun initTodoBody(): RequestBody? {

        val requestBody = ProductTodoSave()
        val detail = ArrayList<ProductTodoSave.DetailsBean>()

        // TODO 获取用户选择的仓库信息
        val houseId = SharedP.getWarehouseId(this@ProductTodoDetailActivity)

//        val warehouseList = userInfo.warehouseList

        // 仓库信息
        val warehouseId: String?
        val warehouseNo: String?
        val warehouseName: String?

        warehouseId = warehouseList[houseId].warehouseId
        warehouseNo = warehouseList[houseId].warehouseNo
        warehouseName = warehouseList[houseId].warehouseName

        val size = oldList.size
        for (i in 0 until size) {
            val view = rv_in_storage_detail.getChildAt(i)

            val boxQrCode = carCodeArray.get(i)

            val bulkNum = view.findViewById<EditText>(R.id.et_box_num).text.toString().trim { it <= ' ' }
            val txt = view.findViewById<EditText>(R.id.et_bulk_num).text.toString().trim { it <= ' ' }
            val startCode = view.findViewById<EditText>(R.id.et_start_code).text.toString().trim { it <= ' ' }
            val endCode = view.findViewById<EditText>(R.id.et_end_code).text.toString().trim { it <= ' ' }
            val sap = view.findViewById<EditText>(R.id.et_sap).text.toString().trim { it <= ' ' }

            val id = positionId.get(i)

            val bean = ProductTodoSave.DetailsBean()
            if (bulkNum.isNotEmpty() && !boxQrCode.isNullOrEmpty()) {

                val goodsId = oldList[i].goodsId

                val subContent = oldList[i].content
                val mergeContent = JsonStringMerge().StringMerge(subContent, content)

                bean.positionId = id
                bean.number = bulkNum
                bean.content = mergeContent

                bean.sapMaterialBatchNo = sap
                bean.startQrCode = startCode
                bean.endQrCode = endCode

                bean.goodsId = goodsId
//                bean.warehouseId = oldList[i].warehouseId
//                bean.warehouseNo = oldList[i].warehouseNo
                bean.warehouseNo = warehouseNo
                bean.goodsNo = oldList[i].goodsNo
                bean.goodsName = oldList[i].goodsName
                bean.boxQrCode = boxQrCode
                bean.unit = oldList[i].unit

                detail.add(bean)
            }
        }
        // 没有合法的操作数据,不请求网络
        if (detail.size == 0) {
            ZBUiUtils.showWarning("请完善您要入库的信息")
            return null
        }

        val remark = et_desc.text.toString().trim { it <= ' ' }
        val time = tv_time.text.toString().trim { it <= ' ' }
        val productDate = product_date.text.toString().trim { it <= ' ' }
        val productSap = product_sap.text.toString().trim { it <= ' ' }

        requestBody.list = detail
//        requestBody.warehouseId = warehouseId
//        requestBody.warehouseNo = warehouseNo
//        requestBody.warehouseName = warehouseName
        requestBody.sapOrderNo = sapOrderNo
        requestBody.moveType = "101"
        requestBody.productionDate = productDate
        requestBody.productionBatchNo = productSap
        requestBody.inTime = time
        requestBody.remark = remark

        val json = DataUtils.toJson(requestBody, 1)
        return RequestBodyJson.requestBody(json)
    }


    override fun showProduct(list: MutableList<ProductDetails.ListBean>) {

        oldList = list
//        tv_house.text = warehouseList[0].warehouseName
        for (i in 0 until warehouseList.size){
            if (warehouseList[i].warehouseNo == list[0].warehouseNo) {
                tv_house.text = warehouseList[i].warehouseName
                SharedP.putWarehouseId(this, i)
                break
            }else{
                if (i == warehouseList.size - 1){
                    ZBUiUtils.showWarning("该物料仓库异常,请选择入库仓库")
                }
            }
        }

        if (adapter == null) {
            val manager = LinearLayoutManager(ResourceUtil.getContext())
            rv_in_storage_detail.addItemDecoration(SpaceItemDecoration(ResourceUtil.dip2px(CodeConstant.ITEM_DECORATION), false))
            rv_in_storage_detail.layoutManager = manager
            adapter = ProductTodoDetailAdapter(this, R.layout.item_produce_detail_todo, oldList)
            rv_in_storage_detail.adapter = adapter

            adapter?.setOnItemClickListener(object : MultiItemTypeAdapter.OnItemClickListener {
                override fun onItemClick(view: View, holder: RecyclerView.ViewHolder, position: Int) {

                    val rlDetail = holder.itemView.findViewById<View>(R.id.rl_detail)
                    val ivArrow = holder.itemView.findViewById<ImageView>(R.id.iv_arrow)
                    val bindingCode = holder.itemView.findViewById<Button>(R.id.btn_binding_code)

                    when (rlDetail.visibility) {
                        View.VISIBLE -> {
                            rlDetail.visibility = View.GONE
                            ivArrow.setImageResource(R.mipmap.ic_collapse)
                        }
                        View.GONE -> {
                            rlDetail.visibility = View.VISIBLE
                            ivArrow.setImageResource(R.mipmap.ic_expand)
                        }
                    }

                    bindingCode.setOnClickListener {
                        itemId = position
                        val intent = Intent(it.context, ScanBoxCodeActivity::class.java)
                        intent.putExtra("itemId", itemId)
                        intent.putExtra("goodsNo", list[position].goodsNo)
                        intent.putExtra(CodeConstant.PAGE_STATE, true)
                        intent.putStringArrayListExtra("boxCodeList", carCodeArray.get(itemId))
                        startActivityForResult(intent, REQUEST_SCAN_CODE)
                    }

                    ZBUiUtils.hideInputWindow(view.context, view)

                }

                override fun onItemLongClick(view: View, holder: RecyclerView.ViewHolder, position: Int): Boolean {
                    return false
                }
            })
        }

    }

    private var dialog: LoadingDialog? = null
    override fun showLoading() {
        dialog = ShowDialog.showFullDialog(this@ProductTodoDetailActivity, "保存中")
    }

    override fun hideLoading() {
        dialog?.close()
    }

    override fun saveSuccess() {
        finish()
    }

    override fun showError(msg: String?) {
        ZBUiUtils.showError(msg)
    }


    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_SCAN_CODE && resultCode == RESULT_SCAN_CODE) {
            itemId = data!!.getIntExtra("itemId", -1)
            boxCodeList = data.getStringArrayListExtra("boxCodeList")
            carCodeArray.put(itemId, boxCodeList)
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        presenter.dispose()
    }

    companion object {

        private const val REQUEST_SCAN_CODE = 1
        private const val RESULT_SCAN_CODE = 2
    }
}
